src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs

```diff
@@ -40,6 +48,216 @@ protected TypeProvider(string defaultNamespace, SourceInputModel? sourceInputMod
+                    break;
+            }
+            if (methodSymbol.IsStatic)
+            {
+                modifiers |= MethodSignatureModifiers.Static;
+            }
+            if (methodSymbol.IsAsync)
+            {
+                modifiers |= MethodSignatureModifiers.Async;
+            }
+            if (methodSymbol.IsVirtual)
+            {
+                modifiers |= MethodSignatureModifiers.Virtual;
+            }
+            if (methodSymbol.IsOverride)
+            {
+                modifiers |= MethodSignatureModifiers.Override;
+            }
+            return modifiers;
+        }
+
+        private static (IList<MethodSignature> Missing, IList<(List<MethodSignature> Current, MethodSignature Previous)> Updated)? CompareMethods(IEnumerable<MethodSignature> currentMethods, IEnumerable<MethodSignature>? previousMethods)
+        {
+            if (previousMethods is null)
+            {
+                return null;
+            }
+            var missing = new List<MethodSignature>();
+            var updated = new List<(List<MethodSignature> Current, MethodSignature Previous)>();
+            var set = currentMethods.ToHashSet(new MethodSignatureComparer());
+            var dict = new Dictionary<string, List<MethodSignature>>();
+            foreach (var item in currentMethods)
+            {
+                if (!dict.TryGetValue(item.Name, out var list))
+                {
+                    dict.Add(item.Name, new List<MethodSignature> { item });
+                }
+                else
+                {
+                    list.Add(item);
+                }
+            }
+            foreach (var item in previousMethods)
+            {
+                if (!set.Contains(item))
+                {
+                    if (dict.TryGetValue(item.Name, out var currentOverloadMethods))
+                    {
+                        updated.Add((currentOverloadMethods, item));
+                    }
+                    else
+                    {
+                        missing.Add(item);
+                    }
+                }
+            }
+            return (missing, updated);
+        }
+
+        private bool TryGetPreviousMethodWithLessOptionalParameters(IList<MethodSignature> currentMethods, MethodSignature previousMethod, [NotNullWhen(true)] out MethodSignature? currentMethodToCall, [NotNullWhen(true)] out IReadOnlyList<Parameter>? missingParameters)
+        {
+            foreach (var item in currentMethods)
+            {
+                if (item.Parameters.Count <= previousMethod.Parameters.Count)
+                {
+                    continue;
+                }
+
+                if (!CurrentContainAllPreviousParameters(previousMethod, item))
+                {
+                    continue;
+                }
+
+                if (!previousMethod.ReturnType.EqualsBySystemType(item.ReturnType))
```

Comment:
based on the context here, I believe we should use `Equals` here because we are trying the see if these return types are the same or not, and `CSharpType.Equals` is written for that purpose.